### ISSUE TITLE
[Design] Possible form updates & enhancements 

### ISSUE THREAD

**Issue by aagonzales** [carbon-design-system/carbon#3167]
From Slack feedback: 

https://ibm-cloudplatform.slack.com/archives/C0M053VPT/p1561408016320400
(the thread was on going, would be good to look at it in full)

![image](https://user-images.githubusercontent.com/11670886/60053412-6a5ca880-969d-11e9-844d-e86f33671aa5.png)
![image](https://user-images.githubusercontent.com/11670886/60053463-82ccc300-969d-11e9-8569-c44f4103b12f.png)

From accessibility channel:
https://ibm-cloudplatform.slack.com/archives/C036P1CTN/p1561410302116200?thread_ts=1561407910.115900&cid=C036P1CTN
![image](https://user-images.githubusercontent.com/11670886/60054024-020ec680-969f-11e9-961d-9b07cb8168d9.png)
https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&showtechniques=331#qr-consistent-behavior-consistent-functionality

**Things to look into:**
- form validation 
   - button states
   - error message
   - Dynamic validation vs on submit guidelines (are they the same or different)
- Interactive state diagram flow
- Optional vs required (include both? but give guidance on when to use which)

**Find examples:** best practice, industry standards, accessibility violations, competitive analysis

**Comment by r4d** [comment]
From the WCAG site:

Examples
A user attempts to submit a form but has neglected to provide input or select a choice in one or more mandatory fields. Using client-side validation, the omission is detected and an alert dialog appears informing the user that mandatory fields have not been completed. The labels of the fields with this problem are changed to identify the problem field, and links to the problem fields are inserted in the document after the submit button so the user can move to them after dismissing the alert.

https://www.w3.org/WAI/WCAG21/Techniques/general/G83.html

A user can't "attempt to submit a form" when the submit button is disabled.

**Comment by vpicone** [comment]
Just to ensure we're having the correct conversation, there's nothing blocking users from implementing forms with both behaviors. The discussion is just around [this gif](https://www.carbondesignsystem.com/components/form/usage#form-validation). There's no text content recommending this pattern, but developers would (rightly) see this and assume that's the way forms should work. 

From researching various large site's login forms (presumably the simplest and most reasonable candidates of this pattern and the component shown in the gif), I haven't been able to track down any examples of disabled submit buttons that become enabled when validation passes. 

By far the most common pattern I've seen is to simply trigger all of the form elements validation on submit. [nytimes](https://myaccount.nytimes.com/auth/login?), github, facebook, google all do this for their login forms. I can see only advantages for following suit in this gif as our primary recommendation for form validation during submission.

But again, developers and designers are currently empowered to do whichever pattern they deem appropriate and accessible for their application. As we only offer individual form elements, and intentionally don't publish a holistic/opinionated form solution (ala [formik](https://github.com/jaredpalmer/formik)) It would be impossible for us to mandate this pattern through code.

**Comment by vpicone** [comment]
For an example of an accessible form element with a disabled primary button, which becomes clickable only when valid, you need only look at the comment box at the bottom of this page.
